### DISCUSSION TITLE
Non overlapping reads support

### DISCUSSION THREAD

**Discussion by gregdenay** [bio-raum/FooDMe2#76] [General]
Hi @marchoeppner 

some colleagues that are actively working on method developement have barcodes that are a bit to long, meaning read merging is not possible. This is in particular true for the published fish barcodes and some of the plants, already while using the 2x250bp chemistry.
Their strategy for now is to run the analysis independently for R1  and R2 and try to manually merge the end results by hand.

I would assume that one could simply merge R1 and R2 after denoising or prior to identity clustering (eventually with a padding sequence in the middle). Since BLAST is a local aligner, opening gaps is not a problem and the successive steps should actually work fine, mybe with some adapted parameters in the blast search.

What do you think?

Implementation-wise it should actually be straigthforward to do it as an experimental feature. VSearch supports this with the `--fastq_join` parameter (instead of `--fastq_merge`) and DADA's `mergePairs` has a `justConcatenate=TRUE` argument. Both insert an `NNN` padding sequence (length 8 for VSearch and 10 for DADA2).

**Comment by marchoeppner** [comment]
Hi,

funnily enough I had a visit from the MRI last week and we discussed the exact same thing, so good timing there ;)

Solution wise, I think this should work.

/M

**Comment by marchoeppner** [comment]
After looking into this some more, I am trying to sum up the "state of things".

- There are certain applications (ITS) that will potentially produce
a) very long amplicons that may exceed the combined read length (non-overlapping)
b) a combination of amplicons below and above the combined read lengths (overlapping and non-overlapping, in the same library). 

Concerning a), we could advise users to not sequence amplicons that are longer than the combined read lengths. But this may not always be possible, given the lengths variability of ITS regions  (our test data includes an amplicon with ~800bp). There is now an option in FooDMe2 (--non_overlapping) to concatenate reads rather than merge them. But this will create unpredictable problems in Blasting, since some of the default parameters will no longer work correctly (% query coverage, specifically). Users will have to adjust this themeselves.

With b), there is currently no clear path forward, other than telling people not do run such experiments. Some of this can be addressed by using longer reads, to not have to join reads at all - within limits (see above). I am also not 100% sure what happens if an amplicon is significantly shorter than the individual read length (should work, right...).  Otherwise, ee can potentially implement a combined merge+join approach, in vsearch (not sure this is even possible in Dada2). Here, reads would first be merged, and all non-merged reads get joined to create mix of merged and joined reads. I have tested this a little bit and it does seem to work, although it adds a level of complexity that could create unforseen issues in some edge cases. 

Right so far?

If so, it may be worthwhile discussing some more with people that run such experiments to understand how much they can do with their designs versus how much additional "complexity" we should add to deal with whatever else that cannot be solved on the design level?

**Reply by gregdenay** [reply]
That's right

It should be clear to users that quality problems will potentially not be detected with either of a) and b) options, and that they should think carefully about the experiment beforehand.
Technically the implemntation you proposed should work. 
it should also possible to do something similar in dada. the `mergePairs` function as an option to keep unmerged reads in the results dataframe, it's then a question of how easy they are to extract to re-run them with the concatenation option.

Checking potential ways to limit problems, I think we should talk about FASTP, I'll open an issue for that

**Comment by gregdenay** [comment]
I'd suggest letting us some time to think about this and maybe implement this in a smaller 1.2 release, eventually along the FASTP changes in Q1 2025

**Comment by gregdenay** [comment]
I just noticed that the VSEARCH workflow either joins or merges, while the DADA workflow either merges or first tries to merge and joins the rest.

Should we rethink this? Shouldn't the mixed approach be an explicit argument? Or shoudl the mixed approach just not exist?
We could change the `--non_overlapping` flag to a `--join_strategy`argument for the 2.0.0 release with options `merge`, `concat`, `flex` or something like this.

**Reply by marchoeppner** [reply]
Is this the current implementation? Checking the vsearch workflow, I see:

```NEXTFLOW

if (params.non_overlapping) {
// First, all mergable reads are merged
        VSEARCH_FASTQMERGE_FOR_JOIN(
            ch_trimmed_reads.paired.map { m, r -> [m, r[0], r[1]] }
        )
        ch_merged_reads = VSEARCH_FASTQMERGE_FOR_JOIN.out.fastq

        // then all unmergable reads are joined
        VSEARCH_FASTQJOIN(
            VSEARCH_FASTQMERGE_FOR_JOIN.out.unmerged_reads
        )

        ch_joined_reads = VSEARCH_FASTQJOIN.out.fastq

       ch_merged_reads
        .mix(ch_joined_reads)
        .collectFile { meta, fastq ->
            [ "${meta.sample_id}.combined.fastq", fastq ]
        }.map { fastq ->
            [ fastq.simpleName ,fastq]
        }.set { ch_all_reads }
...
```

which first merged, and then passes all the unmerged reads to a joining - and then combines each output. 

Sounds about right, no?

**Reply by marchoeppner** [reply]
that aside, yes, this whole aspect is super confusing at the moment. I would simply wish that people only used overlapping paired-end reads.

My thought at the moment:

Yes, provide three options as you suggested, but... determine them based on the data (i.e. insert size distribution vs amplicon size) automatically by default, but allow users to overwrite that automatic selection if they have reason to do so?

Else we only add more confusing command line options that people won't understand?

**Reply by gregdenay** [reply]
Ah yes this seems correct then. In dada the logic is handled in the modules so I didn't thought of checking the subworkflow, I need to reaclimate myself to nextflow ...

I like your idea but how do we get the amplicon size for non overlapping reads, apart for yet another user input. Additionaly the amplicon size can wildly vary between species for some targets.
We could also set a % of reads that should be merged or the pipeline will automatically fallback to concatenate?
Or we handle it properly in the satndard methods and assume users will read the doc when developing new methods?

Not sure we need a concatenate only mode tho, since if using `non_overlapping` it will resort to concatenating anyways? We have to update the doc tho (happy to do it) since this behavior is not described.

**Reply by marchoeppner** [reply]
I think the amplicon min and max size is more or less mandatory information anyway, else the filtering won't work properly, right?

That said, my guess is also that it makes little difference if we request concatenation or not since, as you say, that is already implied with non_overlapping. Question is just if the non_overlapping flag could be set automatically... hm. 

As an alternative, maybe we can capture the fraction of merged reads under normal operations and alert users that the rate of unmerged reads is high and that they should consider using --non_overlapping?

**Reply by gregdenay** [reply]
```
As an alternative, maybe we can capture the fraction of merged reads under normal operations and alert users that the rate of unmerged reads is high and that they should consider using --non_overlapping?
```

I think setting a warning in this case is a sensible option. We are already doing it for the trimming.
Warning users is imo always better than taking the choice for them.

**Reply by marchoeppner** [reply]
Alright. This requires to actually count the number of reads for the merged and unmerged read files. Not entirely sure how I would do that to be honest, without adding bulk to the workflow (like a process just for read counting...). 

FIle size on the other hand is easy, sincew we can just use built-in groovy methods for that. Not terribly precise tho. And for Dada2, we probably need to write something entirely different since everything happens in R, I presume.

**Reply by gregdenay** [reply]
Where should the warning be encoded? In the Groovy process (subworkflow) or in the modules?
I don't think this is a priority tho, we can open an issue and hold in on pause for now time to think about it.

**Reply by marchoeppner** [reply]
i'd put that in the process, after the merging step when --non_overlapping is not set

But yes, this can probably wait. Will add an issue.

**Comment by marchoeppner** [comment]
Technical question... we currently stitch the fwd and rev reads together using an N linker (9 or 10bp). 

With our "optimised" expected min/max size parameters (based on the BarBeQue amplicon length metrics for each primer set +/- twice the std deviation) , I expect that certain systems and read lengths will basically fail the read filtering since the "joined" fragments are a lot shorter than the expected size range. 

I stumbled over this by accident when we did a test sequencing on a v2 micro flow cell of the ring trial samples (most amplicons > 400bp). With only 2x150 + the linker, the data falls way short of the target range. In a sense, this should not be done in production anyway, but since the idea behind the ` --non_overlapping`  option is exactly a situation like this (amplicons too long), I am not sure how we could address this sensibly. 

Thoughts? Solutions? ;)

**Reply by gregdenay** [reply]
I did not think about that. and it makes sense to keep the filtering range within the real biologial range for the BLAST stitching. 
It is always an option to add a parameter that filters based on a min **pseudo-reads** length.  I think the important question though  is where and why do we filter based on length in this case.  On top of my head: 

- should we report BLAST hits where only onbe half of the read has a match or discard them (talking about non-overlapping only here)? If we discard these half-hits then we arguably do not need to filter by merged size prior. Arguably this is already done by setting the `min_coverage` param.
- Does filtering pseudo-reads by size makes sense at all when we don't merge? We already filter (trimmed-)reads by size. 
 
I would tend to skip size filtering when using `--non_overlapping` .

**Reply by marchoeppner** [reply]
Well the filtering in question concerns the final amplicon size (amplicon_min_size, amplicon_max_size). 

Sounds complicated to omit filtering pre-BLAST and only apply filtering after BLAST. Uhm...need to think about that. I would like to avoid a ton of if/else constructs.

**Reply by gregdenay** [reply]
In the `--non_overlapping` workflow `amplicon_min_size` and `amplicon_max_size` are used to define the possible merging range to stitch BLAST hits together after BLAST. Since hits outside of this range are only rported as "half-hits" (so covering only oine halt of the query) their coverage is <50% and they are usually not reported. This could be considered the only real size filtering possiblity for the amplicon size.  In the overlapping workflow, using pseudo-reads is the way to go.

So we could add an if/else in the config of the filtering step to include or omit the size filters? 

I can vaguely remember talking about replacing the DADA2 filtering step by VSEARCH, can you remember why we considered that?

**Reply by gregdenay** [reply]
I just checked the filtersize steps really only filter by size so we could as well just if/else skip them in the subworkflows.

**Reply by marchoeppner** [reply]
I would prefer to solve it like this using the module config(s):

For DADA2_FILTERSIZE:
```
ext.args = [
            params.non_overlapping ? "minLen=50, maxLen=1000" : "maxLen=${params.amplicon_max_length}, minLen=${params.amplicon_min_length}"
        ].join(" ")
```

which would set the min/max length for amplicons to 50 and 1000 respectively, so basically disabling it without omitting the processing step. Would that be OK with you?

**Reply by gregdenay** [reply]
works for me, probably easier for the reporting too.

**Reply by marchoeppner** [reply]
That was the thinking; and as far as I can tell there is also a conversion from fastq to fasta happening in the Vsearch workflow at this point. So just omitting the filtering step would require us to insert an alternative conversion without filtering.

**Reply by marchoeppner** [reply]
Done. There is still a little hickup with the filtering/merging of the BLAST hits where the sizes are also used. For me, that results in everything getting filtered out. But I don't understand the specifics well enough. Maybe best if you just keep working on that bit for your plant data and we can revisit it with other kinds of data later. Else I am happy to share a dataset to reproduce the failure.

**Reply by gregdenay** [reply]
I was actually done with it (or so I thought). Would be great if you could send me some data and the config you used

**Reply by gregdenay** [reply]
Q: fastp finds and remover adapters, shouldn't we disable this? Some of the OTUs end up suspisciously small
 
The problem seems to be that the right tail doesn't appear in the BLAST results:

```
OTU_1;size=360321	KU168685.1_19596	3.79e-60	100.000	231	KU168685.1_19596	13489	125	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168694.1_19594	3.79e-60	100.000	231	KU168694.1_19594	13489	125	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168698.1_19568	3.79e-60	100.000	231	KU168698.1_19568	13489	125	0	0	Dicentrarchus labrax
OTU_1;size=360321	KJ632951.1_19595	1.36e-59	100.000	230	KJ632951.1_19595	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168688.1_19591	1.36e-59	100.000	230	KU168688.1_19591	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168693.1_19590	1.36e-59	100.000	230	KU168693.1_19590	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	EF427552.1_19582	1.36e-59	100.000	230	EF427552.1_19582	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KJ168065.1_19581	1.36e-59	100.000	230	KJ168065.1_19581	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	AF143189.1_19580	1.36e-59	100.000	230	AF143189.1_19580	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	X81566.1_19579	1.36e-59	100.000	230	X81566.1_19579	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KJ686347.1_19578	1.36e-59	100.000	230	KJ686347.1_19578	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	EU036430.1_19577	1.36e-59	100.000	230	EU036430.1_19577	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	EU036429.1_19576	1.36e-59	100.000	230	EU036429.1_19576	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	LC384919.1_19573	1.36e-59	100.000	230	LC384919.1_19573	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168700.1_19571	1.36e-59	100.000	230	KU168700.1_19571	13489	124	0	0	Dicentrarchus labrax
OTU_1;size=360321	KU168699.1_19598	1.76e-58	99.200	226	KU168699.1_19598	13489	125	1	0	Dicentrarchus labrax
OTU_1;size=360321	KU168695.1_19592	1.76e-58	99.200	226	KU168695.1_19592	13489	125	1	0	Dicentrarchus labrax
OTU_1;size=360321	KU168697.1_19570	1.76e-58	99.200	226	KU168697.1_19570	13489	125	1	0	Dicentrarchus labrax
OTU_1;size=360321	KU168686.1_19569	1.76e-58	99.200	226	KU168686.1_19569	13489	125	1	0	Dicentrarchus labrax
OTU_1;size=360321	KU168691.1_19593	2.95e-56	98.387	219	KU168691.1_19593	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	AF143190.1_19588	2.95e-56	98.387	219	AF143190.1_19588	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	EF427553.1_19587	2.95e-56	98.387	219	EF427553.1_19587	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	KJ168064.1_19586	2.95e-56	98.387	219	KJ168064.1_19586	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	AF165071.1_19585	2.95e-56	98.387	219	AF165071.1_19585	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	EU492155.1_19584	2.95e-56	98.387	219	EU492155.1_19584	13489	124	2	0	Dicentrarchus labrax
OTU_1;size=360321	KU168696.1_19575	3.82e-55	97.600	215	KU168696.1_19575	13489	125	3	0	Dicentrarchus labrax
```

Compare with a manual BLAST online. From the online results one half always seems to align kind of crappy. 
Relaxing the blast search parameters (`--blast_qcov 80 --blast_perc_id 80`) seems to fix it. I pulled these values out my hat, so you may want to be inspect them carefully. This leads to pairs of hits being found for references sequences. The range seems off to stitch them tho, but relaxing that too (very crude test with 100-600) solves it:

```
            "cluster_ids": "OTU_1[1.0]; OTU_6[1.0]; OTU_13[1.0]; OTU_16[1.0]; OTU_21[1.0]; OTU_22[1.0]",
            "name": "Dicentrarchus labrax",
            "proportion": 0.9985,
            "rank": "species",
            "reads": 360714,
            "taxid": "13489"
```

Not sure how much of this is due to the aggressive adapter trimming and how much due to crappy database match though.

**Reply by marchoeppner** [reply]
> Q: fastp finds and remover adapters, shouldn't we disable this? Some of the OTUs end up suspisciously small

Hm, never looked into it much.  I suppose one may want to disable it (we do for the first fastp trimming, I think, using `-A`. 
To avoid any risk of fastp recognizing primer sites as adapters, we could add `-A` to `--fastp_options` . That said, the actual fastp trimming step comes after cutadapt, so should at least not create an issue with primer sites. Bot not 100% sure how fastp would deal with low complexity data like amplicons.

**Reply by gregdenay** [reply]
I added the `-A` option to fastp. The tool detects adapter by aligning reads and detecting conserved sequences, def a problem in metabarcoding... 
Also a performance hit.
